pkg/unikontainers/ipc_test.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2024 Nubificus LTD.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package unikontainers
+
+import (
+	"net"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestGetSockAddr(t *testing.T) {
+	dir := "/tmp"
+	name := "test.sock"
+	expected := filepath.Join(dir, name)
+	result := getSockAddr(dir, name)
+	assert.Equal(t, expected, result, "Expected %s, but got %s", expected, result)
+}
+
+func TestGetInitSockAddr(t *testing.T) {
+	containerDir := "/tmp/container"
+	expected := filepath.Join(containerDir, initSock)
+	result := getInitSockAddr(containerDir)
+	assert.Equal(t, expected, result, "Expected %s, but got %s", expected, result)
+}
+
+func TestGetUruncSockAddr(t *testing.T) {
+	containerDir := "/tmp/container"
+	expected := filepath.Join(containerDir, uruncSock)
+	result := getUruncSockAddr(containerDir)
+	assert.Equal(t, expected, result, "Expected %s, but got %s", expected, result)
+}
+
+func TestEnsureValidSockAddr(t *testing.T) {
+	validSockAddr := "/tmp/valid.sock"
+	emptySockAddr := ""
+	longSockAddr := string(make([]byte, 109))
+
+	assert.NoError(t, ensureValidSockAddr(validSockAddr), "Expected no error for valid socket address")
+	assert.Error(t, ensureValidSockAddr(emptySockAddr), "Expected error for empty socket address")
+	assert.Error(t, ensureValidSockAddr(longSockAddr), "Expected error for long socket address")
+}
+
+func TestSockAddrExists(t *testing.T) {
+	existingSockAddr := "/tmp/existing.sock"
+	nonExistingSockAddr := "/tmp/non_existing.sock"
+
+	// Create a temporary socket file
+	f, err := os.Create(existingSockAddr)
+	if err != nil {
+		t.Fatalf("Failed to create temporary socket file: %v", err)
+	}
+	defer os.Remove(existingSockAddr)
+	f.Close()
+
+	assert.True(t, SockAddrExists(existingSockAddr), "Expected socket address to exist")
+	assert.False(t, SockAddrExists(nonExistingSockAddr), "Expected socket address to not exist")
+}
+
+func TestSendIPCMessage(t *testing.T) {
+	socketAddress := "/tmp/test.sock"
+	message := ReexecStarted
+
+	listener, err := net.Listen("unix", socketAddress)
+	if err != nil {
+		t.Fatalf("Failed to create listener: %v", err)
+	}
+	defer os.Remove(socketAddress)
+	defer listener.Close()
```

Comment:
Should we first close the listener and then remove the socket?  Or because of defer it does not matter?

Reply:
The defer statements are executed last in, first out. So, the listener is closed first and then the socket is removed.

Reply:
Oh ok, no need for change then.